### PR TITLE
Support for Common Lisp

### PR DESCRIPTION
**Pull request by theHamsta**
Since everything is very WIP I don't plan to merge this any time soon since it requires really a lot more work and I need to make some decision how stuff is supposed to work. But suggestions are always welcome.

The parser is an extension of @sogaiu 's tree-sitter-clojure with some rules overwritten.

~~TODO: use `@symbol`~~

### CODE REVIEW

**Comment by Happy-Dude** [comment]
@theHamsta this is exciting; I'm actually super happy to see something like this!

I've always found [vim-sexp](https://github.com/guns/vim-sexp) and [vim-sexp-mappings-for-regular-people](https://github.com/tpope/vim-sexp-mappings-for-regular-people) incredibly powerful; unfortunately in most vim-startup benchmarks, it is probably one of the biggest culprits in slow vim startup time.

Was wondering if there was any consideration in making a generalized [S-expression parser](https://www.cs.unm.edu/~luger/ai-final2/LISP/CH%2011_S-expressions,%20The%20Syntax%20of%20Lisp.pdf) and extending it further for LISP languages like Common Lisp and Clojure?

**Comment by theHamsta** [comment]
I guess there is already a s-expr parser (https://github.com/AbstractMachinesLab/tree-sitter-sexp).

I've used this code in the early days of nvim-treesitter for Lisps https://github.com/theHamsta/crazy-node-movement . I really works good for manipulating and moving within sexpr. Now it moved away to other languages. But implementing the functions of paredit-like functions is really easy. However, it requires some effort to bundle the code and maintaining it. Let's see whether someone finds time for it. Maybe I find some time to look into it.

**Comment by Happy-Dude** [comment]
Awesome; thanks for the context.

@theHamsta I realize that several nvim-treesitter folks are maintaining a bulk of the parsers supported by the plugin. Is there a sponsorship structure (maybe through GitHub) that can be set up to support y'alls work with maintaining tree-sitter on neovim?

**Comment by theHamsta** [comment]
@Happy-Dude did you test the highlights/locals already?

**Comment by Happy-Dude** [comment]
> @Happy-Dude did you test the highlights/locals already and have some feedback (rather for the approach than for details)? I guess the whole built-in function/marco highlighting has to wait until `any-of?` is merged in Neovim. Things like smart rename should already work. All the rest of the parsing (number literals, 1+, etc is still very clojure but is usually not a problem for most source files)
> 
> 

I have not tested; I commented because it was relevant as I wanted to learn Common Lisp some more. If you would like me to, I can play around with the branch on my side while I work on some exercises.

**Comment by theHamsta** [comment]
Maybe for a beginner the default highlighting is better since it knows all functions defined by the standard. But would be nice when you experiment with it😉

**Comment by theHamsta** [comment]
I created a specialized plugin that adds proper highlighting for functions/macros in the `cl` namespace

**Comment by theHamsta** [comment]
@sogaiu this will change clojure keywords to the symbol highlight group we could change `TSSymbol` to the `Type` highlight from `Identifier` to have it what is was before and what it has in standard Vim highlighting.

```
  hi def link lispKey			Type
```

File: queries/commonlisp/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,142 @@
+(sym_lit) @variable
+
+;; A highlighting for functions/macros in th cl namespace is available in theHamsta/nvim-treesitter-commonlisp
+;(list_lit . (sym_lit) @function.builtin (#cl-standard-symbol? @function.builtin))
```
Comment (theHamsta) [inline queries/commonlisp/highlights.scm:4]: the plugin is basically only two tables with the standard symbols of Common Lisp

File: queries/commonlisp/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ .
+ (sym_lit) @function
+ (#eq? @function.macro "deftest"))
+
+;;; Macros and Special Operators
+(list_lit
+ .
+ (sym_lit) @function.macro
+ ;; For a complete and more efficient version install theHamsta/nvim-treesitter-commonlisp
+ (#match? @function.macro
```
Comment (stsewd) [inline queries/commonlisp/highlights.scm:114]: believe this is any-of

```suggestion
 (#any-of? @function.macro
```

**Review by stsewd — APPROVED** [review]

**Pull request merged**